src/antidote_crdt_secure_counter_pn.erl

```diff
@@ -0,0 +1,173 @@
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either expressed or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% List of the contributors to the development of Antidote: see AUTHORS file.
+%% Description and complete License: see LICENSE file.
+%% -------------------------------------------------------------------
+
+%% antidote_crdt_secure_counter_pn: A convergent, replicated, operation
+%% based secure (using the Paillier cryptosystem) PN-Counter.
+
+-module(antidote_crdt_secure_counter_pn).
+
+-behaviour(antidote_crdt).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([
+    new/0,
+    value/1,
+    downstream/2,
+    update/2,
+    equal/2,
+    to_binary/1,
+    from_binary/1,
+    is_operation/1,
+    require_state_downstream/1
+]).
+
+-type freshness() :: fresh | spoiled.
+-type delta() :: integer().
+-type nsquare() :: integer().
+-type state() :: {freshness(), integer(), nsquare()}.
+-type op() :: {increment, delta()} | {increment, {delta(), nsquare()}}.
+-type effect() :: delta() | {delta(), nsquare()}.
+
+%% @doc Create a new, empty 'antidote_crdt_secure_counter_pn'
+-spec new() -> state().
+new() ->
+    {fresh, 0, 0}.
+
+%% @doc The single, total value of a secure pn-counter.
+-spec value(state()) -> integer().
+value({_, Value, _}) when is_integer(Value) ->
+    Value.
+
+%% @doc Generate a downstream operation.
+%% The first parameter is either a tuple of the form `{increment, integer()}'
+%% or a tuple of the form `{increment, {integer(), integer()}}'. The first
+%% integer represents the delta value by which the counter will be incremented,
+%% while the second integer is the n squared value used by the Paillier
+%% cryptosystem. The second parameter is the secure pn-counter, this parameter
+%% is not actually used.
+-spec downstream(op(), state()) -> {ok, effect()}.
+downstream({increment, {Delta, NSquare}}, _SecurePNCounter)
+    when is_integer(Delta) and is_integer(Delta)
+->
+    {ok, {Delta, NSquare}};
+downstream({increment, Delta}, _SecurePNCounter) when is_integer(Delta) ->
+    {ok, Delta}.
+
+%% @doc Updates a given secure pn-counter, incrementing it by a given `Delta'.
+%% By incrementing we mean performing the homomorphic addition of the counter
+%% value with the given delta. As described by the Paillier cryptosystem, the
+%% homomorphic addition of plaintexts translates to the product of two
+%% ciphertexts modulo n squared.
+%%
+%% A value for `NSquare` may also be provided, if it isn't, the one in the
+%% counter state will be used.
+%%
+%% Returns the updated secure pn-counter.
+-spec update(effect(), state()) -> {ok, state()}.
+update({Delta, NSquare}, {fresh, _, _}) ->
+    {ok, {spoiled, Delta, NSquare}};
+update({Delta, NSquare}, {spoiled, Value, _}) ->
```

Comment:
```suggestion
update({Delta, NSquare}, {spoiled, Value, NSquare}) ->
```
Does it make sense to change the N² value in a later increment? If not, make it an error.